src/main/java/com/fasterxml/jackson/databind/ObjectMapper.java

```diff
@@ -2277,6 +2277,7 @@ public ObjectMapper setConstructorDetector(ConstructorDetector cd) {
     public ObjectMapper setCacheProvider(CacheProvider cacheProvider) {
         _deserializationConfig = _deserializationConfig.with(cacheProvider);
         _serializationConfig = _serializationConfig.with(cacheProvider);
+        _deserializationContext = _deserializationContext.with(cacheProvider.forDeserializerCache(_deserializationConfig));
```

Comment:
Actually... since we will probably have multiple caches being re-created, I think it'd be better to have a `withCaches(CacheProvider)` method instead -- then `DefaultDeserializationContext` would handle instantiation and re-construction. And we would only need to add one method in there.